.github/workflows/ci.yml

```diff
@@ -64,16 +63,43 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: ['3.7', '3.8', '3.9', '3.10']
+        python: ['3.7', '3.8', '3.9', '3.10', '3.11']
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
       - name: Setup python
-        uses: actions/setup-python@v2
+        uses: actions/setup-python@v4
         with:
           python-version: '${{ matrix.python }}'
       - name: Run tests
         run: ./ci.sh
-        env:
-          # Should match 'name:' up above
-          JOB_NAME: 'macOS (${{ matrix.python }})'
+      - name: "Upload coverage data"
+        uses: "actions/upload-artifact@v3"
+        with:
+          name: coverage-data
+          path: "empty/.coverage.*"
+          if-no-files-found: error
+
+  coverage:
+    if: always()
```

Comment:
nit: Is this `if` necessary?

Reply:
Yes! I even plan to add it to urllib3 if you agree. This ensures that the coverage step runs even if some jobs failed (but still waiting for everything to finish).

In urllib3 sometimes one single job fails because of a flaky test and we have no idea if coverage is still at 100% or not. The test suite will still be failing, but that way we don't need to wait for everything to pass to know if coverage is still at 100%. It will also cover the case when as admins we just merge a change even if unrelated tests are failing: we have merged uncovered code in the past like that.

Reply:
Ooooo I love that!! TIL! Thanks Quentin!